runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java

```diff
@@ -0,0 +1,136 @@
+  private final ControlClientSource clientSource;
+
+  private DockerEnvironmentFactory(
+      DockerWrapper docker,
+      GrpcFnServer<FnApiControlClientPoolService> controlServiceServer,
+      GrpcFnServer<GrpcLoggingService> loggingServiceServer,
+      GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer,
+      GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer,
+      Supplier<String> idGenerator,
+      ControlClientSource clientSource) {
+    this.docker = docker;
+    this.controlServiceServer = controlServiceServer;
+    this.loggingServiceServer = loggingServiceServer;
+    this.retrievalServiceServer = retrievalServiceServer;
+    this.provisioningServiceServer = provisioningServiceServer;
+    this.idGenerator = idGenerator;
+    this.clientSource = clientSource;
+  }
+
+  /** Creates a new, active {@link RemoteEnvironment} backed by a local Docker container. */
+  @Override
+  public RemoteEnvironment createEnvironment(Environment environment) throws Exception {
+    String workerId = idGenerator.get();
+
+    // Prepare docker invocation.
+    Path workerPersistentDirectory = Files.createTempDirectory("worker_persistent_directory");
+    Path semiPersistentDirectory = Files.createTempDirectory("semi_persistent_dir");
+    String containerImage = environment.getUrl();
+    // TODO: https://issues.apache.org/jira/browse/BEAM-4148 The default service address will not
+    // work for Docker for Mac.
+    String loggingEndpoint = loggingServiceServer.getApiServiceDescriptor().getUrl();
+    String artifactEndpoint = retrievalServiceServer.getApiServiceDescriptor().getUrl();
+    String provisionEndpoint = provisioningServiceServer.getApiServiceDescriptor().getUrl();
+    String controlEndpoint = controlServiceServer.getApiServiceDescriptor().getUrl();
+    List<String> args =
+        Arrays.asList(
+            "-v",
+            // TODO: Mac only allows temporary mounts under /tmp by default (as of 17.12).
+            String.format("%s:%s", workerPersistentDirectory, semiPersistentDirectory),
+            // NOTE: Host networking does not work on Mac, but the command line flag is accepted.
+            "--network=host",
+            containerImage,
+            String.format("--id=%s", workerId),
+            String.format("--logging_endpoint=%s", loggingEndpoint),
+            String.format("--artifact_endpoint=%s", artifactEndpoint),
+            String.format("--provision_endpoint=%s", provisionEndpoint),
+            String.format("--control_endpoint=%s", controlEndpoint),
+            String.format("--semi_persist_dir=%s", semiPersistentDirectory));
+
+    // Wrap the blocking call to clientSource.get in case an exception is thrown.
+    String containerId = null;
+    InstructionRequestHandler instructionHandler;
+    try {
+      containerId = docker.runImage(containerImage, args);
+      // Wait on a client from the gRPC server.
+      instructionHandler = clientSource.get(workerId);
```

Comment:
<!--thread_id:cc_182929276_t; commit:fb7d2bd7ba1bb27186e8d2dda861acd08666a0c1; resolved:1-->
<!--section:context-quote-->
> **angoenka** wrote:
> It seems that we are piggy backing on control channel to see when worker comes up.
> 1. Should we have a timeout for container startup and retry?
> 2. Should we use some docker specific signaling to know when client is up?

<!--section:body-->
1) This is part of the container contract and is outside of the purview of this specific change. Unfortunately, containers may in general take arbitrarily long to boot up, so we don't have a reasonable starting point as of now. @herohde would have more insight. Note that I did make the client pool itself use configurable timeouts when waiting for incoming clients.

2) The only signal docker can give us is whether a container is _not_ running. Just because the container has booted and is running does not mean the SDK harness is in a healthy state. Incoming control client requests are the only way we can positively confirm that SDK harnesses are running. Note that we do plan to include follow-up work (as part of this same bug) to allow early failure signaling if docker knows that a container is no longer running.

Reply:
As @bsidhom mentions, a container may take an arbitrary amount of time to start up (installing stuff from a slow internet, etc) and we have no good way to know what is happening between container start (= process start) and it connects to the control api (= it's fully initialized and ready to do work). It's up to each runner to impose a timeout, if desired. Dataflow fails a job if there is no progress for a couple of hours, for example.

The runner is responsible for bringing up the container (using docker directly, via Kubernetes, etc) and should restart it, when needed. Most runners are probably going to largely defer to some docker management technology for handling this aspect.

Reply:
Given the docker limitation of exposing process health, this seems reasonable.

Reply:
However, I want to put this in a get-with-timeout loop which logs every n minutes, just to make sure we have at least a bit of information about "we tried lots and it didn't come up"

Reply:
That's a good idea. We could do that within DockerEnvironmentFactory and use an independent executor to induce timeouts on the blocking call (somewhat ugly and wasteful), we could change the client source to always take a timeout parameter (annoying in tests and other places where we don't want timeouts), or we could augment client source with another method and make it no longer a functional interface (annoying in this case but possibly the best approach.

This actually makes me wonder if it's worth returning a future from the client source and making this explicit and configurable by clients. That does come with the cost of making use of the source annoying because we will want the client to be materialized pretty much everywhere. Preferences?

Reply:
I've gone ahead and changed `ControlClientPool.Soure.get` to take a timeout parameter and added logging. Let me know what you think and I'll rebase this into the commit history.